### DISCUSSION TITLE
Save writable pdf as readonly pdf

### DISCUSSION THREAD

**Discussion by Zapato92** [stephanrauh/ngx-extended-pdf-viewer#1356] [Q&A] [answered]
Hi,

Is there any way to save an editable pdf with a form as a readonly pdf ? 

Until now I managed to save the pdf with the data entered but the produced pdf is still editable.
I also tried to save each page (with getPageAsImage() ) and reinsert them into a new PDF but the images do not contain the data entered in the form.

Thanks!

**Comment by stephanrauh** [comment, marked as answer]
That's out of scope of this library, but in theory, it should be simple. Basically, you only have to add `0x40` to the annotation flags of each field annotation in the form. Unfortunately, this isn't persistent if you do this with pdf.js.

I also looked for third-party libraries. There are several libraries out there modifying PDF files, but none of them cover your use case. Incredible!

The only option I've found it using iText. You can read a PDF file with iText, call the method `flattenForms()`, and save it again. If you tech stack includes C# or Java, iText is your friend! https://stackoverflow.com/questions/48120555/editable-pdf-to-non-editable-read-only-after-filling-in-the-acro-fields-using-it

There's also https://www.npmjs.com/package/pdf-flatten, but it seems to do the trick by converting every page to an image. If that's true, that's probably not what you want to do. Plus, this library hasn't seen maintenance since three years, so I wouldn't base a productive project on it.

**Reply by Zapato92** [reply]
Thanks for your answer and for taking the time to study my use case :)

I will take a look at the iText solution which seems to be the most appropriate for the moment.

And a big thank you for all the work you do on your library.

**Reply by stephanrauh** [reply]
I'm always glad to help! Oh, and most of the time, working on the library is fun, not work... but recently, the popularity of the library has grown beyond my wildest dreams, and so has the number of bug tickets. You don't happen to have some spare time to share, contributing to a successful open source project? :)

As for iText: keep me in the loop, please. I'm curious if this approach works, and if it's really as little work as I believe it to be.
